engine/telemetry/pubsub.go

```diff
@@ -0,0 +1,826 @@
+}
+
+// clientsFor returns the relevant client IDs for a given span, traversing
+// through parents, in random order.
+func (ps *PubSub) clientsFor(traceID trace.TraceID, spanID trace.SpanID) []string {
+	ps.spansL.Lock()
+	defer ps.spansL.Unlock()
+	key := spanKey{
+		TraceID: traceID,
+		SpanID:  spanID,
+	}
+	// slog := slog.With("traceID", traceID, "spanID", spanID)
+	clients := map[string]struct{}{}
+	for {
+		if client, ok := ps.spanClients[key]; ok {
+			// slog.ExtraDebug("!!! ClientsFor found client", "client", client, "keySpan", key.SpanID)
+			clients[client] = struct{}{}
+		}
+		if parent, ok := ps.spanParents[key]; ok && parent.IsValid() {
+			key.SpanID = parent
+		} else {
+			// slog.ExtraDebug("!!! ClientsFor did not find parent")
+			break
+		}
+	}
+	var ids []string
+	for id := range clients {
+		ids = append(ids, id)
+	}
+	return ids
+}
+
+func (ps *PubSub) trackSpans(spans []sdktrace.ReadOnlySpan) {
+	ps.spansL.Lock()
+	defer ps.spansL.Unlock()
+	for _, span := range spans {
+		ps.trackSpan(span)
+	}
+}
+
+func (ps *PubSub) filterAbandonedSpans(spans map[trace.SpanID]sdktrace.ReadOnlySpan) []sdktrace.ReadOnlySpan {
+	ps.spansL.Lock()
+	defer ps.spansL.Unlock()
+	var kept []sdktrace.ReadOnlySpan
+	for _, span := range spans {
+		if ps.isAbandonedL(span) {
+			slog.ExtraDebug("dropping abandoned span", "span", span.Name(), "spanID", span.SpanContext().SpanID())
+			continue
+		}
+		kept = append(kept, span)
+	}
+	return kept
+}
+
+// isAbandonedL returns true if the span or any of its parents are done.
+func (ps *PubSub) isAbandonedL(span sdktrace.ReadOnlySpan) bool {
+	key := spanKey{
+		TraceID: span.SpanContext().TraceID(),
+		SpanID:  span.SpanContext().SpanID(),
+	}
+	for {
+		if ps.spanDone[key] {
+			return true
+		}
+		if parent, ok := ps.spanParents[key]; ok && parent.IsValid() {
+			key.SpanID = parent
+		} else {
+			return false
+		}
+	}
+}
+
+func (ps *PubSub) trackSpan(span sdktrace.ReadOnlySpan) {
+	key := spanKey{
+		TraceID: span.SpanContext().TraceID(),
+		SpanID:  span.SpanContext().SpanID(),
+	}
+
+	// slog := slog.With("span", span.Name(), "spanID", span.SpanContext().SpanID(), "parentID", span.Parent().SpanID())
+
+	ps.spanParents[key] = span.Parent().SpanID()
+
+	// slog.ExtraDebug("!!! ClientAnnotator tracking span clients")
+
+	for _, attr := range span.Attributes() {
+		if attr.Key == telemetry.ClientIDAttr {
+			// slog.ExtraDebug("!!! > trackSpan tracking client ID", "value", attr.Value.AsString())
+			ps.spanClients[key] = attr.Value.AsString()
+			return
+		}
+	}
+
+	// slog.ExtraDebug("!!! trackSpan did not find client ID")
+}
+
+type ClientMetadata struct {
+	// ClientID is unique to each client. The main client's ID is the empty string,
+	// any module and/or nested exec client's ID is a unique digest.
+	ClientID string `json:"client_id"`
+}
+
+const (
+	// NB: a bit odd to move this here, but feels less risky than duplicating
+	ClientMetadataMetaKey = "x-dagger-client-metadata"
+)
```

Comment:
nit: if we changed the field `ClientMetadata.Labels` in `engine/opts.go` to just be `map[string]string` could we just import that `ClientMetadata` type into this package and use it? Or do we not want to use that type here for other reasons?

I don't care about where the const is defined or anything like that really, just seeing if we can de-dupe some of the code like `decodeMeta` and such.

Reply:
Yeah that unties the knot - done, thanks!